### PR TITLE
Api invalid claim event

### PR DESCRIPTION
**Pull request by porcellus**
## Summary of change
(A few sentences about this PR)

## Related issues
- Link to issue1 here
- Link to issue1 here

## Test Plan
(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes
(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates
- [ ] Changelog has been updated
- [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
   - Along with the associated array in `lib/ts/version.ts`
- [ ] Changes to the version if needed
   - In `package.json`
   - In `package-lock.json`
   - In `lib/ts/version.ts`
- [ ] Had run `npm run build-pretty`
- [ ] Had installed and ran the pre-commit hook
- [ ] Issue this PR against the latest non released version branch.
   - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
   - If no such branch exists, then create one from the latest released branch.

## Remaining TODOs for this PR
- [ ] Item1
- [ ] Item2

### CODE REVIEW

File: lib/ts/types.ts [hunk 1]
```diff
@@ -22,6 +22,10 @@ export type Event =
           action: "SIGN_OUT" | "REFRESH_SESSION" | "SESSION_CREATED" | "ACCESS_TOKEN_PAYLOAD_UPDATED";
           userContext: any;
       }
+    | {
+          action: "API_INVALID_CLAIM";
+          claimValidationError: ClaimValidationError;
```
Comment (rishabhpoddar) [inline lib/ts/types.ts:27]:
```suggestion
          claimValidationError: ClaimValidationError[];
```

File: lib/ts/utils/index.ts [hunk 2]
```diff
@@ -96,6 +96,11 @@ export function validateAndNormaliseInputOrThrowError(options: InputType): Norma
         sessionExpiredStatusCode = options.sessionExpiredStatusCode;
     }
 
+    let invalidClaimStatusCode = 403;
+    if (options.invalidClaimStatusCode !== undefined) {
+        invalidClaimStatusCode = options.invalidClaimStatusCode;
+    }
+
```
Comment (rishabhpoddar) [inline lib/ts/utils/index.ts:103]:
```suggestion
if (invalidClaimStatusCode === sessionExpiredStatusCode) {
   throw some error
}
```

File: lib/ts/fetch.ts [hunk 3]
```diff
@@ -298,6 +298,9 @@ export default class AuthHttpRequest {
                     }
                     logDebugMessage("doRequest: Retrying original request");
                 } else {
+                    if (response.status === AuthHttpRequest.config.invalidClaimStatusCode) {
+                        onInvalidClaimResponse(JSON.parse(response.headers.get("invalid-claim")!));
```
Comment (rishabhpoddar) [inline lib/ts/fetch.ts:302]: - We should extract `JSON.parse(response.headers.get("invalid-claim")!)` into a function. We can call this function `getInvalidClaimsFromResponse`. This should be compatible with fetch and axios response.
- We decided that we will not be sending invalid-claims as a header, and instead will be sending it as a body.

File: lib/ts/axios.ts [hunk 4]
```diff
@@ -190,6 +191,9 @@ export function responseInterceptor(axiosInstance: any) {
                     true
                 );
```
Comment (rishabhpoddar) [inline lib/ts/axios.ts:192]: Add tests

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/axios.ts (hunk 4 above)
Comment (porcellus) [reply lib/ts/axios.ts:192]: I'll add tests in a separate PR with both #136 and this merged.

File: lib/ts/recipeImplementation.ts [hunk 5]
```diff
@@ -137,6 +144,17 @@ export default function RecipeImplementation(recipeImplInput: {
             }
 
             // we do not send an event here since it's triggered in setIdRefreshToken area.
+        },
+
+        async getInvalidClaimsFromResponse(response: AxiosResponse | Response): Promise<ClaimValidationError[]> {
```
Comment (rishabhpoddar) [inline lib/ts/recipeImplementation.ts:149]: Can you make the function signature of this similar to other functions in this file?

File: lib/ts/axios.ts [hunk 6]
```diff
@@ -245,6 +251,12 @@ export function responseErrorInterceptor(axiosInstance: any) {
                 true
             );
         } else {
+            if (
+                error.response !== undefined &&
+                error.response.status === AuthHttpRequestFetch.config.invalidClaimStatusCode
+            ) {
+                onInvalidClaimResponse(JSON.parse(error.response.headers["invalid-claim"]));
```
Comment (rishabhpoddar) [inline lib/ts/axios.ts:258]:
```suggestion
                onInvalidClaimResponse(await AuthHttpRequestFetch.recipeImpl.getInvalidClaimsFromResponse(response));
```

File: lib/ts/axios.ts [hunk 7]
```diff
@@ -399,6 +411,9 @@ export default class AuthHttpRequest {
                         }
                         logDebugMessage("doRequest: Retrying original request");
                     } else {
+                        if (response.status === AuthHttpRequestFetch.config.invalidClaimStatusCode) {
+                            onInvalidClaimResponse(JSON.parse(response.headers["invalid-claim"]));
```
Comment (rishabhpoddar) [inline lib/ts/axios.ts:415]:
```suggestion
                            onInvalidClaimResponse(await AuthHttpRequestFetch.recipeImpl.getInvalidClaimsFromResponse(response));
```

File: lib/ts/axios.ts [hunk 8]
```diff
@@ -436,6 +452,9 @@ export default class AuthHttpRequest {
                             }
                             logDebugMessage("doRequest: Retrying original request");
                         } else {
+                            if (response.status === AuthHttpRequestFetch.config.invalidClaimStatusCode) {
+                                onInvalidClaimResponse(JSON.parse(response.headers["invalid-claim"]));
```
Comment (rishabhpoddar) [inline lib/ts/axios.ts:456]:
```suggestion
                                onInvalidClaimResponse(await AuthHttpRequestFetch.recipeImpl.getInvalidClaimsFromResponse(response));
```

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipeImplementation.ts (hunk 5 above)
Comment (porcellus) [reply lib/ts/recipeImplementation.ts:149]: I can, but a `userContext` doesn't make sense here, it'd be always undefined and not reused.

Comment (rishabhpoddar) [reply lib/ts/recipeImplementation.ts]: Also need to expose this function via the index file so that the user can use it

**Review by rishabhpoddar — APPROVED** [review]

**Pull request merged**